src/core/src/op/transpose.cpp

```diff
@@ -95,20 +95,41 @@ bool Transpose::evaluate(TensorVector& outputs, const TensorVector& inputs) cons
             int4_extract_t m_half;
         };
 
-        auto out_ptr = int4_iterator(static_cast<uint8_t*>(out.data()));
-        // The int4_iterator not supports const pointer but these data are not modified
-        auto in_ptr = int4_iterator(static_cast<uint8_t*>(const_cast<void*>(arg.data())));
-        if ((arg_type == ov::element::i4 || arg_type == ov::element::u4) && arg.get_shape().size() == 2) {
-            const auto out_shape_d0 = out_shape[0];
-            const auto out_shape_d1 = out_shape[1];
-
-            for (size_t i = 0; i < out_shape_d0; i++) {
-                size_t off = i;
-                for (size_t j = 0; j < out_shape_d1; j++) {
-                    out_ptr.copy_from(in_ptr + off);
-                    ++out_ptr;
-                    off += out_shape_d0;
+        if (arg_type == ov::element::i4 || arg_type == ov::element::u4) {
+            // The int4_iterator not supports const pointer but these data are not modified
+            auto transpose_xy =
+                [] (int4_iterator& out_ptr, int4_iterator& in_ptr, size_t out_shape_d0, size_t out_shape_d1) {
+                    for (size_t i = 0; i < out_shape_d0; i++) {
+                        size_t off = i;
+                        for (size_t j = 0; j < out_shape_d1; j++) {
+                            out_ptr.copy_from(in_ptr + off);
+                            ++out_ptr;
+                            off += out_shape_d0;
+                        }
                 }
+            };
+            if (arg.get_shape().size() == 2) {
+                auto out_ptr = int4_iterator(static_cast<uint8_t*>(out.data()));
+                auto in_ptr = int4_iterator(static_cast<uint8_t*>(const_cast<void*>(arg.data())));
+                const auto out_shape_d0 = out_shape[0];
+                const auto out_shape_d1 = out_shape[1];
+                transpose_xy(out_ptr, in_ptr, out_shape_d0, out_shape_d1);
+            } else if (arg.get_shape().size() == 3) {
+                OPENVINO_ASSERT(axes_order[0] == 0, "Transpose for i4/u4 dtype is supported only for byx=>bxy case");
```

Comment:
Why it has to be changed when it is used e.g. is there any constant folding etc?
Here is generic implementation  and information `byx=>bxy` should not be there. 
Limit the i4/u4 implementation to same rank, shapes order.

Make proper implementation of this type in reference and use element::iterator which allow jump over elements in common way and here just call this reference at some conditions (as limited support)

Add unit test for this values in `src/plugins/template/tests/functional/op_reference/transpose.cpp`

Reply:
Yes, make follow up ticket to correct it.
The previous changes was also done like WA and and refactoring was not done. it means to add some kind of new reference implementation for LP.
The current version can be also extracted as reference implementation

Reply:
Sure lets make a ticket to update the reference part. 
Do you have any preferred bucket?